File: android/app/src/main/java/com/targetzone/library/ui/auth/AuthViewModel.kt
package com.targetzone.library.ui.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.targetzone.library.data.model.User
import com.targetzone.library.data.repository.AuthRepository
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

data class AuthState(
    val isLoading: Boolean = false,
    val error: String? = null,
    val otpSent: Boolean = false,
    val otpVerified: Boolean = false,
    val sessionToken: String? = null,
    val isNewUser: Boolean = false,
    val isLoggedIn: Boolean = false,
    val user: User? = null,
    // OTP resend / SMS-fallback timing — mirrors the web LoginPage and the
    // backend's 10s resend cooldown (AuthService.checkCooldownAndGenerateOtp).
    // otpSendCount tracks how many times an OTP has been (re)sent;
    // secondsSinceSend counts up to 10 after each send, driving both "Resend
    // OTP" availability and the "Send via SMS instead" offer (shown once a
    // resend has also gone 10s without verification).
    val otpSendCount: Int = 0,
    val secondsSinceSend: Int = 0,
    val smsOptionUsed: Boolean = false
)

class AuthViewModel(private val repo: AuthRepository) : ViewModel() {
    private val _state = MutableStateFlow(AuthState())
    val state: StateFlow<AuthState> = _state

    private var resendTimerJob: Job? = null

    private fun startResendTimer() {
        resendTimerJob?.cancel()
        _state.value = _state.value.copy(secondsSinceSend = 0)
        resendTimerJob = viewModelScope.launch {
            while (_state.value.secondsSinceSend < 10) {
                delay(1000)
                _state.value = _state.value.copy(secondsSinceSend = _state.value.secondsSinceSend + 1)
            }
        }
    }

    fun sendOtp(contact: String, contactType: String = "MOBILE") = viewModelScope.launch {
        _state.value = _state.value.copy(isLoading = true, error = null)
        repo.sendOtp(contact, contactType)
            .onSuccess {
                _state.value = _state.value.copy(isLoading = false, otpSent = true, otpSendCount = 1)
                startResendTimer()
            }
            .onFailure { _state.value = _state.value.copy(isLoading = false, error = it.message) }
    }

    fun resendOtp(contact: String, contactType: String = "MOBILE") = viewModelScope.launch {
        _state.value = _state.value.copy(isLoading = true, error = null)
        repo.sendOtp(contact, contactType)
            .onSuccess {
                _state.value = _state.value.copy(isLoading = false, otpSendCount = _state.value.otpSendCount + 1)
                startResendTimer()
            }
            .onFailure { _state.value = _state.value.copy(isLoading = false, error = it.message) }
    }

    fun sendOtpViaSms(contact: String, contactType: String = "MOBILE") = viewModelScope.launch {
        _state.value = _state.value.copy(isLoading = true, error = null)
        repo.sendOtp(contact, contactType, channel = "SMS")
            .onSuccess {
                _state.value = _state.value.copy(isLoading = false, smsOptionUsed = true, otpSendCount = _state.value.otpSendCount + 1)
                startResendTimer()
            }
            .onFailure { _state.value = _state.value.copy(isLoading = false, error = it.message) }
    }

    fun verifyOtp(mobile: String, otp: String) = viewModelScope.launch {
        _state.value = _state.value.copy(isLoading = true, error = null)
        repo.verifyOtp(mobile, otp)
            .onSuccess { res ->
                _state.value = _state.value.copy(
                    isLoading = false, otpVerified = true,
                    sessionToken = res.sessionToken, isNewUser = res.newUser
                )
            }
            .onFailure { _state.value = _state.value.copy(isLoading = false, error = it.message) }
    }

    fun register(name: String, email: String?, sessionToken: String, dateOfBirth: String? = null, gender: String? = null, address: String? = null) = viewModelScope.launch {
        _state.value = _state.value.copy(isLoading = true, error = null)
        repo.register(name, email, sessionToken, dateOfBirth, gender, address)
            .onSuccess { auth -> _state.value = _state.value.copy(isLoading = false, isLoggedIn = true, user = auth.user) }
            .onFailure { _state.value = _state.value.copy(isLoading = false, error = it.message) }
    }

    fun login(sessionToken: String) = viewModelScope.launch {
        _state.value = _state.value.copy(isLoading = true, error = null)
        repo.login(sessionToken)
            .onSuccess { auth -> _state.value = _state.value.copy(isLoading = false, isLoggedIn = true, user = auth.user) }
            .onFailure { _state.value = _state.value.copy(isLoading = false, error = it.message) }
    }

    fun adminLogin(contact: String, otp: String) = viewModelScope.launch {
        _state.value = _state.value.copy(isLoading = true, error = null)
        repo.adminLogin(contact, otp)
            .onSuccess { auth -> _state.value = _state.value.copy(isLoading = false, isLoggedIn = true, user = auth.user) }
            .onFailure { _state.value = _state.value.copy(isLoading = false, error = it.message) }
    }

    fun resetOtpState() {
        resendTimerJob?.cancel()
        _state.value = AuthState()
    }
    fun clearError() { _state.value = _state.value.copy(error = null) }
}

File: ios/LibrarySystem/LibrarySystem/Features/Auth/AuthViewModel.swift
import Foundation
import Combine

@MainActor
final class AuthViewModel: ObservableObject {
    @Published var isLoading = false
    @Published var error: String?
    @Published var otpSent = false
    @Published var sessionToken: String?
    @Published var isNewUser = false

    // OTP resend / SMS-fallback timing — mirrors the web LoginPage and the
    // backend's 10s resend cooldown (AuthService.checkCooldownAndGenerateOtp).
    // otpSendCount tracks how many times an OTP has been (re)sent;
    // secondsSinceSend counts up to 10 after each send, driving both "Resend
    // OTP" availability and the "Send via SMS instead" offer (shown once a
    // resend has also gone 10s without verification).
    @Published var otpSendCount = 0
    @Published var secondsSinceSend = 0
    @Published var smsOptionUsed = false
    private var resendTimerTask: Task<Void, Never>?

    var canResend: Bool { secondsSinceSend >= 10 }
    var secondsLeft: Int { max(0, 10 - secondsSinceSend) }
    // SMS fallback only makes sense for a phone contact — mirrors the web
    // LoginPage's showSmsOption, which gates the same way.
    var showSmsOption: Bool { contactType == "MOBILE" && !smsOptionUsed && otpSendCount >= 2 && canResend }

    private let repo = AuthRepository.shared
    private let tokenManager = TokenManager.shared

    var contact = ""
    var contactType = "MOBILE"   // "MOBILE" | "EMAIL" — mirrors the web LoginPage's toggle

    func sendOtp(contact: String, contactType: String = "MOBILE") {
        self.contact = contact
        self.contactType = contactType
        isLoading = true; error = nil
        Task {
            do {
                try await repo.sendOtp(contact: contact, contactType: contactType)
                otpSent = true
                otpSendCount = 1
                startResendTimer()
            } catch { self.error = error.localizedDescription }
            isLoading = false
        }
    }

    func resendOtp() {
        isLoading = true; error = nil
        Task {
            do {
                try await repo.sendOtp(contact: contact, contactType: contactType)
                otpSendCount += 1
                startResendTimer()
            } catch { self.error = error.localizedDescription }
            isLoading = false
        }
    }

    func sendOtpViaSms() {
        isLoading = true; error = nil
        Task {
            do {
                try await repo.sendOtp(contact: contact, contactType: contactType, channel: "SMS")
                smsOptionUsed = true
                otpSendCount += 1
                startResendTimer()
            } catch { self.error = error.localizedDescription }
            isLoading = false
        }
    }

    private func startResendTimer() {
        resendTimerTask?.cancel()
        secondsSinceSend = 0
        resendTimerTask = Task { [weak self] in
            while let self = self, self.secondsSinceSend < 10 {
                try? await Task.sleep(nanoseconds: 1_000_000_000)
                if Task.isCancelled { return }
                self.secondsSinceSend += 1
            }
        }
    }

    func verifyOtp(otp: String) {
        isLoading = true; error = nil
        Task {
            do {
                let resp = try await repo.verifyOtp(contact: contact, otp: otp)
                sessionToken = resp.sessionToken
                isNewUser = resp.newUser
                if !resp.newUser { await login(sessionToken: resp.sessionToken) }
            } catch { self.error = error.localizedDescription }
            isLoading = false
        }
    }

    func register(name: String, email: String?, dateOfBirth: String?,
                  gender: String?, address: String?) {
        guard let token = sessionToken else { return }
        isLoading = true; error = nil
        Task {
            do {
                let auth = try await repo.register(name: name, email: email,
                                                   sessionToken: token, dateOfBirth: dateOfBirth,
                                                   gender: gender, address: address)
                tokenManager.save(token: auth.accessToken, user: auth.user)
            } catch { self.error = error.localizedDescription }
            isLoading = false
        }
    }

    private func login(sessionToken: String) async {
        do {
            let auth = try await repo.login(sessionToken: sessionToken)
            tokenManager.save(token: auth.accessToken, user: auth.user)
        } catch { self.error = error.localizedDescription }
    }

    func sendAdminOtp(contact: String) {
        self.contact = contact
        isLoading = true; error = nil
        Task {
            do {
                try await repo.sendOtp(contact: contact)
                otpSent = true
                otpSendCount = 1
                startResendTimer()
            } catch { self.error = error.localizedDescription }
            isLoading = false
        }
    }

    func adminLogin(otp: String) {
        isLoading = true; error = nil
        Task {
            do {
                let auth = try await repo.adminLogin(contact: contact, otp: otp)
                tokenManager.save(token: auth.accessToken, user: auth.user)
            } catch { self.error = error.localizedDescription }
            isLoading = false
        }
    }

    func resetOtpState() {
        resendTimerTask?.cancel()
        otpSent = false
        sessionToken = nil
        isNewUser = false
        error = nil
        contact = ""
        contactType = "MOBILE"
        otpSendCount = 0
        secondsSinceSend = 0
        smsOptionUsed = false
    }
}
